alibi_detect/cd/pytorch/mmd_online.py

```diff
@@ -0,0 +1,191 @@
+            data_type: Optional[str] = None
+    ) -> None:
+        """
+        Online maximum Mean Discrepancy (MMD) data drift detector using preconfigured thresholds.
+
+        Parameters
+        ----------
+        x_ref
+            Data used as reference distribution.
+        ert
+            The expected run-time (ERT) in the absence of drift.
+        window_size
+            The size of the sliding test-window used to compute the test-statistic.
+            Smaller windows focus on responding quickly to severe drift, larger windows focus on
+            ability to detect slight drift.
+        preprocess_x_ref
+            Whether to already preprocess and store the reference data.
+        preprocess_fn
+            Function to preprocess the data before computing the data drift metrics.
+        kernel
+            Kernel used for the MMD computation, defaults to Gaussian RBF kernel.
+        sigma
+            Optionally set the GaussianRBF kernel bandwidth. Can also pass multiple bandwidth values as an array.
+            The kernel evaluation is then averaged over those bandwidths.
+        n_bootstraps
+            The number of bootstrap simulations used to configure the thresholds. The larger this is the
+            more accurately the desired ERT will be targeted. Should ideally be at least an order of magnitude
+            larger than the ERT.
+        device
+            Device type used. The default None tries to use the GPU and falls back on CPU if needed.
+            Can be specified by passing either 'cuda', 'gpu' or 'cpu'. Only relevant for 'pytorch' backend.
+        input_shape
+            Shape of input data.
+        data_type
+            Optionally specify the data type (tabular, image or time-series). Added to metadata.
+        """
+        super().__init__(
+            x_ref=x_ref,
+            ert=ert,
+            window_size=window_size,
+            preprocess_x_ref=preprocess_x_ref,
+            preprocess_fn=preprocess_fn,
+            sigma=sigma,
+            n_bootstraps=n_bootstraps,
+            input_shape=input_shape,
+            data_type=data_type
+        )
+        self.meta.update({'backend': 'pytorch'})
+
+        # set backend
+        if device is None or device.lower() in ['gpu', 'cuda']:
+            self.device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
+            if self.device.type == 'cpu':
+                print('No GPU detected, fall back on CPU.')
+        else:
+            self.device = torch.device('cpu')
+
+        # initialize kernel
+        sigma = torch.from_numpy(sigma).to(self.device) if isinstance(sigma, np.ndarray) else None
+        self.kernel = kernel(sigma) if kernel == GaussianRBF else kernel
+
+        # compute kernel matrix for the reference data
+        self.x_ref = torch.from_numpy(self.x_ref).to(self.device)
+        self.k_xx = self.kernel(self.x_ref, self.x_ref, infer_sigma=(sigma is None))
+        self.infer_sigma = False
```

Comment:
`self.infer_sigma` redundant since `sigma` either given or already inferred from reference data?

Reply:
So if I understand correctly, `sigma` will never be inferred as is *traditionally* done via the median distance heuristic between samples x and y?